include/xtensor/xhistogram.hpp

```diff
@@ -0,0 +1,312 @@
+     * @return An one-dimensional xarray<double>, length: bin_edges.size()-1.
+     */
+    template<class E1, class E2>
+    inline auto histogram(
+        E1&& data, E2&& bin_edges, bool density=false
+    )
+    {
+        using value_type = typename std::decay_t<E1>::value_type;
+
+        return histogram(data, bin_edges, xt::ones<value_type>({data.size()}), density);
+    }
+
+    /**
+     * @ingroup histogram
+     * @brief Compute the histogram of a set of data.
+     *
+     * @param data The data.
+     * @param bins The number of bins. [default: 10]
+     * @param density If true the resulting integral is normalized to 1. [default: false]
+     * @return An one-dimensional xarray<double>, length: bin_edges.size()-1.
+     */
+    template<class E1>
+    inline auto histogram(
+        E1&& data, size_t bins=10, bool density=false
+    )
+    {
+        using value_type = typename std::decay_t<E1>::value_type;
+
+        auto bin_edges = histogram_bin_edges(data, xt::ones<value_type>({data.size()}), bins);
+
+        return histogram(data, bin_edges, xt::ones<value_type>({data.size()}), density);
+    }
+
+    /**
+     * @ingroup histogram
+     * @brief Compute the histogram of a set of data.
+     *
+     * @param data The data.
+     * @param bins The number of bins.
+     * @param weights Weight factors corresponding to each data-point.
+     * @param density If true the resulting integral is normalized to 1. [default: false]
+     * @return An one-dimensional xarray<double>, length: bin_edges.size()-1.
+     */
+    template<class E1, class E2>
+    inline auto histogram(
+        E1&& data, size_t bins, E2&& weights, bool density=false
+    )
+    {
+        auto bin_edges = histogram_bin_edges(data, weights, bins);
+
+        return histogram(data, bin_edges, weights, density);
+    }
+
+    /**
+     * @ingroup histogram
+     * @brief Defines different algorithms in "histogram_bin_edges"
+     */
+    enum class histogram_algorithm { automatic, linspace, uniform };
+
+    /**
+     * @ingroup histogram
+     * @brief Compute the bin-edges of a histogram of a set of data using different algorithms.
+     *
+     * @param data The data.
+     * @param weights Weight factors corresponding to each data-point.
+     * @param left The lower-most edge.
+     * @param right The upper-most edge.
+     * @param bins The number of bins. [default: 10]
+     * @param mode The type of algorithm to use. [default: "auto"]
+     * @return An one-dimensional xarray<double>, length: bins+1.
+     */
+    template<class E1, class E2, class E3>
+    inline auto histogram_bin_edges(const E1 &data, const E2 &weights,
+        E3 left, E3 right, size_t bins=10, histogram_algorithm mode=histogram_algorithm::automatic
+    )
+    {
+        // counter and return type
+        using size_type = detail::common_size_type_t<E1,E2>;
+        using value_type = typename E1::value_type;
+        using weights_type = typename E2::value_type;
```

Comment:
@JohanMabille By what should I replace ``using size_type = detail::common_size_type_t<E1,E2>;`` in that case? My compiler complains:

```cpp
In file included from /data/prog/forks/xtensor/include/xtensor/xoperation.hpp:19:
/data/prog/forks/xtensor/include/xtensor/xfunction.hpp:45:54: error: type 'const
      xt::xtensor_container<xt::uvector<double, std::__1::allocator<double> >, 1, xt::layout_type::row_major,
      xt::xtensor_expression_tag> &' cannot be used prior to '::' because it has no members
            using type = std::common_type_t<typename Args::size_type...>;
                                                     ^
/data/prog/forks/xtensor/include/xtensor/xfunction.hpp:55:9: note: in instantiation of template class
      'xt::detail::common_size_type<const xt::xtensor_container<xt::uvector<double, std::__1::allocator<double> >, 1,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &, const xt::xbroadcast<xt::xscalar<double>,
      std::__1::array<unsigned long, 1> > &>' requested here
        using common_size_type_t = typename common_size_type<Args...>::type;
        ^
/data/prog/forks/xtensor/include/xtensor/xhistogram.hpp:178:35: note: in instantiation of template type alias
      'common_size_type_t' requested here
        using size_type = detail::common_size_type_t<E1,E2>;
```

Reply:
```cpp
using size_type = detail::common_size_type_t<std::decay_t<E1>, std::decay_t<E2>>;
```
should fix the problem.

Whenever you take a universal reference, you need to "decay" it to access its inner typedefs.